### PR TITLE
feat(noUselessEscapeInRegex): add rule

### PR DESCRIPTION
**Pull request by Conaclos**
## Summary

Implement partially #1090

## Test Plan

I added all the tests of the ESLint rules.
We have a single divergence:  we report `\k` as useless in non unicode-aware regexes.

### CODE REVIEW

File: crates/biome_js_analyze/src/lint/nursery/no_useless_escape_in_regex.rs [hunk 1]
```diff
@@ -0,0 +1,287 @@
...
+                    diag.note("The character must only be escaped when it is the first character of the class.")
+                }
+                b'B' => {
+                    diag.note("The escape sequence has only a meaning outside of a characvter class.")
+                }
+                b'.' | b'$' | b'*' | b'+' | b'?' | b'{' | b'}' | b'[' | b'(' | b')' => {
+                    diag.note("The character must only be escaped when it is outside of a characvter class.")
+                }
+                b'-' => {
+                    diag.note("The character must be escaped only when it appears in the niddle of the character class.")
```
Comment (arendjr) [inline crates/biome_js_analyze/src/lint/nursery/no_useless_escape_in_regex.rs:243]:
```suggestion
                    diag.note("The character should only be escaped when it appears in the middle of the character class.")
```

File: crates/biome_js_analyze/src/lint/nursery/no_useless_escape_in_regex.rs [hunk 2]
```diff
@@ -0,0 +1,287 @@
...
+        } else if *in_char_class {
+            match escaped {
+                b'^' => {
+                    diag.note("The character must only be escaped when it is the first character of the class.")
+                }
+                b'B' => {
+                    diag.note("The escape sequence has only a meaning outside of a characvter class.")
+                }
+                b'.' | b'$' | b'*' | b'+' | b'?' | b'{' | b'}' | b'[' | b'(' | b')' => {
+                    diag.note("The character must only be escaped when it is outside of a characvter class.")
```
Comment (arendjr) [inline crates/biome_js_analyze/src/lint/nursery/no_useless_escape_in_regex.rs:240]:
```suggestion
                    diag.note("The character should only be escaped when it is outside of a character class.")
```

File: crates/biome_js_analyze/src/lint/nursery/no_useless_escape_in_regex.rs [hunk 3]
```diff
@@ -0,0 +1,287 @@
...
+        );
+        Some(if matches!(escaped, b'p' | b'P' | b'k') {
+            diag.note("The escape sequence is only useful when the Regular Expression is unicode-aware. To be unicode-aware, the `u` or `v` flag must be used.")
+        } else if *in_char_class {
+            match escaped {
+                b'^' => {
+                    diag.note("The character must only be escaped when it is the first character of the class.")
+                }
+                b'B' => {
+                    diag.note("The escape sequence has only a meaning outside of a characvter class.")
```
Comment (arendjr) [inline crates/biome_js_analyze/src/lint/nursery/no_useless_escape_in_regex.rs:237]:
```suggestion
                    diag.note("The escape sequence only has meaning outside of a character class.")
```

File: crates/biome_js_analyze/src/lint/nursery/no_useless_escape_in_regex.rs [hunk 4]
```diff
@@ -0,0 +1,287 @@
...
+            markup! {
+                "The character doesn't need to be escaped."
+            },
+        );
+        Some(if matches!(escaped, b'p' | b'P' | b'k') {
+            diag.note("The escape sequence is only useful when the Regular Expression is unicode-aware. To be unicode-aware, the `u` or `v` flag must be used.")
+        } else if *in_char_class {
+            match escaped {
+                b'^' => {
+                    diag.note("The character must only be escaped when it is the first character of the class.")
```
Comment (arendjr) [inline crates/biome_js_analyze/src/lint/nursery/no_useless_escape_in_regex.rs:234]:
```suggestion
                    diag.note("The character should only be escaped when it is the first character of the class.")
```

File: crates/biome_js_analyze/src/lint/nursery/no_useless_escape_in_regex.rs [hunk 5]
```diff
@@ -0,0 +1,287 @@
...
+            .nth(1)?;
+        let diag = RuleDiagnostic::new(
+            rule_category!(),
+            TextRange::at(backslash_position, (1 + *escaped_byte_len as u32).into()),
+            markup! {
+                "The character doesn't need to be escaped."
+            },
+        );
+        Some(if matches!(escaped, b'p' | b'P' | b'k') {
+            diag.note("The escape sequence is only useful when the Regular Expression is unicode-aware. To be unicode-aware, the `u` or `v` flag must be used.")
```
Comment (arendjr) [inline crates/biome_js_analyze/src/lint/nursery/no_useless_escape_in_regex.rs:230]:
```suggestion
            diag.note("The escape sequence is only useful if the Regular Expression is unicode-aware. To be unicode-aware, the `u` or `v` flag should be used.")
```

**Review by arendjr — APPROVED** [review]
Nice one, good diagnostics too! Just some nits on the wording.

File: crates/biome_js_analyze/src/lint/nursery/no_useless_escape_in_regex.rs [hunk 6]
```diff
@@ -0,0 +1,290 @@
...
+use biome_js_syntax::{JsRegexLiteralExpression, JsSyntaxKind, JsSyntaxToken};
+use biome_rowan::{AstNode, BatchMutationExt, TextRange, TextSize};
+
+use crate::JsRuleAction;
+
+declare_lint_rule! {
+    /// Disallow unnecessary escape sequence in regular expression literals.
+    ///
+    /// Escaping non-special characters in regular expression literals doesn't have any effect.
+    /// However, they may confuse a reader.
```
Comment (ematipico) [inline crates/biome_js_analyze/src/lint/nursery/no_useless_escape_in_regex.rs:15]:
```suggestion
    /// Escaping non-special characters in regular expression literals doesn't have any effect. Hence, they may confuse a reader.
```

File: crates/biome_js_analyze/src/lint/nursery/no_useless_escape_in_regex.rs [hunk 7]
```diff
@@ -0,0 +1,290 @@
...
+    fn action(ctx: &RuleContext<Self>, state: &Self::State) -> Option<JsRuleAction> {
+        let State {
+            backslash_index, ..
+        } = state;
+        // Add 1 because the index was computed in the pattern (it doesn't take `/` into account).
+        let adjusted_backslash_index = (*backslash_index as usize) + 1;
+        let node = ctx.query();
+        let value_token = node.value_token().ok()?;
+        let regex_text = value_token.text_trimmed();
+        debug_assert!(regex_text.as_bytes().get(adjusted_backslash_index) == Some(&b'\\'));
```
Comment (ematipico) [inline crates/biome_js_analyze/src/lint/nursery/no_useless_escape_in_regex.rs:264]: Can you provide a second argument? The message should explain why the assertion fails

**Review by ematipico — APPROVED** [review]

**Pull request merged**